apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/RowAsyncEnumerator.cs

```diff
@@ -90,18 +124,39 @@ public RowAsyncEnumerator(BigtableServiceApiClient.ReadRowsStream stream)
 
             IEnumerator<Row> MergeResponseChunks(ReadRowsResponse response)
             {
+                ByteString previouslyProcessedKey = _lastCompletedRowKey;
                 foreach (var chunk in response.Chunks)
                 {
                     _rowMergeState = _rowMergeState.HandleChunk(this, chunk);
 
                     if (chunk.CommitRow)
                     {
                         _rowMergeState = _rowMergeState.CommitRow(this);
-
-                        // TODO: Capture response.LastScannedRowKey here for smart retries
+                        _lastCompletedRowKey = _currentCell.Row.Key;
+                        _requestManager.IncrementRowsReadSoFar();
                         yield return _currentCell.Row;
                     }
                 }
+
+                if (previouslyProcessedKey != _lastCompletedRowKey)
+                {
+                    // There was a full row found in the response.
+                    UpdateLastFoundKey(_lastCompletedRowKey);
+                }
+                else
+                {
+                    // The service indicates that it processed rows that did not match the filter, and will not
```

Comment:
I don't understand the comment - how do we know this from the fact that `_lastCompletedRowKey` is `previouslyProcessedRowKey`? I suspect this is just a bit of the protocol I'm unaware of, but it may be worth expanding the comment a little.

Reply:
This comment was copied from [the Java implementation](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/6d25f6b109302f2d84d473f76c0ef5093f15f1b3/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RetryingReadRowsOperation.java#L191). It looks like it is slightly inaccurate because the `LastScannedRowKey` could be null, but `LastScannedRowKey` (and its analogue in Java) ignore null or empty row keys. I'll update the comment and local method name to be clearer.